Presentation docs selectednodeinstances (#3028)
iTwin/itwinjs-core: docs/presentation/Content/SelectedNodeInstances.md

# Selected Node Instances Specification

> TypeScript type: [SelectedNodeInstancesSpecification]($presentation-common).

Returns content for selected (input) instances.

## Attributes

| Name                                                                | Required? | Type                                                                                | Default |
| ------------------------------------------------------------------- | --------- | ----------------------------------------------------------------------------------- | ------- |
| *Filtering*                                                         |
| [`acceptableSchemaName`](#attribute-acceptableschemaname)           | No        | `string`                                                                            | `""`    |
| [`acceptableClassNames`](#attribute-acceptableclassnames)           | No        | `string[]`                                                                          | `[]`    |
| [`acceptablePolymorphically`](#attribute-acceptablepolymorphically) | No        | `boolean`                                                                           | `false` |

> | [`onlyIfNotHandled`](#attribute-onlyifnothandled)                   | No        | boolean                                                                             | `false` |

Reviewer: ```suggestion
| [`onlyIfNotHandled`](#attribute-onlyifnothandled)                   | No        | `boolean`                                                                           | `false` |
```

| [`relatedProperties`](#attribute-relatedproperties)                 | No        | `RelatedPropertiesSpecification[]`                                                  | `[]`    |
| [`calculatedProperties`](#attribute-calculatedproperties)           | No        | `CalculatedPropertiesSpecification[]`                                               | `[]`    |
| [`propertyCategories`](#attribute-propertycategories)               | No        | `PropertyCategorySpecification[]`                                                   | `[]`    |
| [`propertyOverrides`](#attribute-propertyoverrides)                 | No        | `PropertySpecification[]`                                                           | `[]`    |
| [`showImages`](#attribute-showimages)                               | No        | `boolean`                                                                           | `false` |
| *Misc.*                                                             |
| [`relatedInstances`](#attribute-relatedinstances)                   | No        | [`RelatedInstanceSpecification[]`](../Common-Rules/RelatedInstanceSpecification.md) | `[]`    |

### Attribute: `acceptableSchemaName`

> >> **Default behaviour:** All schema names accepted

Reviewer: ```suggestion
> **Default behaviour:** All schema names accepted
```

Specifies ECSchema name which the input instances have to match for the specification to be used.

```ts
[[include:SelectedNodeInstances.AcceptableSchemaName.Ruleset]]
```

  | Selected input                  | Result                                                                                                          |
  | ------------------------------- | --------------------------------------------------------------------------------------------------------------- |
  | `BisCore.SpatialViewDefinition` | ![Example when selecting "SpatialViewDefinition"](./media/selectednodeinstances-with-spatialviewdefinition.png) |
  | `Generic.GroupModel`            | ![Example when selecting "GroupModel"](./media/content-empty-table.png)                                         |

### Attribute: `acceptableClassNames`

> >> **Default behaviour:** All class names accepted

Reviewer: ```suggestion
> **Default behaviour:** All class names accepted
```

  | ------------------------------- | --------------------------------------------------------------------------------------------------------------- |
  | `BisCore.SpatialViewDefinition` | ![Example when selecting "SpatialViewDefinition"](./media/selectednodeinstances-with-spatialviewdefinition.png) |
  | `BisCore.DictionaryModel`       | ![Example when selecting "DictionaryModel"](./media/content-empty-table.png)                                    |

### Attribute: `acceptablePolymorphically`

> **Default value:** `false`

Specifies whether derived classes of `acceptableClassNames` should be included in the content.

```ts
[[include:SelectedNodeInstances.AcceptablePolymorphically.Ruleset]]
```

  | Selected input                  | Result                                                                                                        |
  | ------------------------------- | ------------------------------------------------------------------------------------------------------------- |
  | `BisCore.DictionaryModel`       | ![Example when selecting "DictionaryModel"](./media/selectednodeinstances-with-acceptablepolymorphically.png) |

>   | `BisCore.SpatialViewDefinition` | ![Example when selecting "SpatialViewDefinition"](./media/content-empty-table.png)                            |

Reviewer: I think a better example would be to have a ruleset with `acceptableClassNames` pointing to a class that has instances of itself and its subclass (not sure if our test imodels have such a case, though). Then we could pass 2 instance keys for the example - one for the class instance and one for the subclass instance. When using `acceptablePolymorphically: true`, content for both instances would be returned. When using `acceptablePolymorphically: false`, only content for the base class instance would be returned.

Author: I tried searching for such case but I did not seem to find one. I did update the example to somewhat match the other content specifications when handling polymorphism.

Reviewer: That'll do, thanks

>   | `true`                                                                           | ![Example when selecting "View Definition" with "acceptable polymorphically"](./media/selectednodeinstances-with-spatialviewdefinition.png) |
>   | `false`                                                                          | ![Example when doing a simple "View Definition" select](./media/content-empty-table.png)                                                    |

Reviewer: ```suggestion
  | `true`                                                                           | ![Example of "acceptable polymorphically" attribute set to "true"](./media/selectednodeinstances-with-spatialviewdefinition.png) |
  | `false`                                                                          | ![Example of "acceptable polymorphically" attribute set to "false"](./media/content-empty-table.png)                                                    |
```
